compiler/src/dotty/tools/MainGenericCompiler.scala

```diff
@@ -0,0 +1,195 @@
+    this.copy(javaProps = javaProps.appendedAll(args.toList))
+
+  def withResidualArgs(args: String*): CompileSettings =
+    this.copy(residualArgs = residualArgs.appendedAll(args.toList.filter(_.nonEmpty)))
+
+  def withScriptArgs(args: String*): CompileSettings =
+    this.copy(scriptArgs = scriptArgs.appendedAll(args.toList.filter(_.nonEmpty)))
+
+  def withTargetScript(file: String): CompileSettings =
+    Try(Source.fromFile(file)).toOption match
+      case Some(_) => this.copy(targetScript = file)
+      case None      =>
+        println(s"not found $file")
+        this.copy(exitCode = 2)
+  end withTargetScript
+
+  def withCompiler: CompileSettings =
+    this.copy(compiler = true)
+
+  def withQuiet: CompileSettings =
+    this.copy(quiet = true)
+
+  def withColors: CompileSettings =
+    this.copy(colors = true)
+
+  def withNoColors: CompileSettings =
+    this.copy(colors = false)
+}
+
+object MainGenericCompiler {
+
+  val classpathSeparator = File.pathSeparator
+
+  @sharable val javaOption = raw"""-J(.*)""".r
+  @sharable val javaPropOption = raw"""-D(.+?)=(.?)""".r
+  @tailrec
+  def process(args: List[String], settings: CompileSettings): CompileSettings = args match
+    case Nil =>
+      settings
+    case "--" :: tail =>
+      process(Nil, settings.withResidualArgs(tail.toList*))
+    case ("-v" | "-verbose" | "--verbose") :: tail =>
+      process(tail, settings.withScalaArgs("-verbose"))
+    case ("-q" | "-quiet") :: tail =>
+      process(tail, settings.withQuiet)
+    case "-Oshort" :: tail =>
+      process(tail, settings.withJavaArgs("-XX:+TieredCompilation", "-XX:TieredStopAtLevel=1"))
+    case "-repl" :: tail =>
+      process(tail, settings.withCompileMode(CompileMode.Repl))
+    case "-script" :: targetScript :: tail =>
+      process(Nil, settings
+        .withCompileMode(CompileMode.Script)
+        .withJavaProps("script.path" -> targetScript)
+        .withTargetScript(targetScript)
+        .withScriptArgs(tail.toList*))
+    case "-compile" :: tail =>
+      process(tail, settings.withCompileMode(CompileMode.Compile))
+    case "-decompile" :: tail =>
+      process(tail, settings.withCompileMode(CompileMode.Decompile))
+    case "-print-tasty" :: tail =>
+      process(tail, settings.withCompileMode(CompileMode.PrintTasty))
+    case "-run" :: tail =>
+      process(tail, settings.withCompileMode(CompileMode.Run))
+    case "-colors" :: tail =>
+      process(tail, settings.withColors)
+    case "-no-colors" :: tail =>
+      process(tail, settings.withNoColors)
+    case "-with-compiler" :: tail =>
+      process(tail, settings.withCompiler)
+    case ("-cp" | "-classpath" | "--class-path") :: cp :: tail =>
+      val cpEntries = cp.split(classpathSeparator).toList
+      val singleEntryClasspath: Boolean = cpEntries.sizeIs == 1
+      val globdir: String = if singleEntryClasspath then cp.replaceAll("[\\\\/][^\\\\/]*$", "") else "" // slash/backslash agnostic
```

Comment:
I don't understand where this treatment of globs comes from. It doesn't seem like the script had something like that.

Reply:
It is copied from the MainGenericRunner, I think it is meant to be a problem that appears in cygwin

Reply:
There are several things going on here.
A valid wildcard classpath entry is vulnerable to a JVM bug.  See [trailing-asterisks-on-windows-jvm-command-line-args-are-globbed-by-jvm](https://stackoverflow.com/questions/37037375/trailing-asterisks-on-windows-jvm-command-line-args-are-globbed-in-cygwin-bash-s)
In Windows, both forward slash and backslash are legal as file.separator, and both forms must be handled.

Reply:
the `globdir` code works around problems caused by a Windows jvm bug.

Reply:
I have moved the code to a shared method between both scripts